Code/Mantid/Framework/Algorithms/src/ExtractSpectra.cpp

```diff
@@ -0,0 +1,494 @@
+      "Name of the output workspace");
+
+  declareProperty("XMin", EMPTY_DBL(), "An X value that is within the first "
+                                       "(lowest X value) bin that will be "
+                                       "retained\n"
+                                       "(default: workspace min)");
+  declareProperty("XMax", EMPTY_DBL(), "An X value that is in the highest X "
+                                       "value bin to be retained (default: max "
+                                       "X)");
+  auto mustBePositive = boost::make_shared<BoundedValidator<int>>();
+  mustBePositive->setLower(0);
+  declareProperty("StartWorkspaceIndex", 0, mustBePositive,
+                  "The index number of the first entry in the Workspace that "
+                  "will be loaded\n"
+                  "(default: first entry in the Workspace)");
+  // As the property takes ownership of the validator pointer, have to take care
+  // to pass in a unique pointer to each property.
+  declareProperty(
+      "EndWorkspaceIndex", EMPTY_INT(), mustBePositive,
+      "The index number of the last entry in the Workspace to be loaded\n"
+      "(default: last entry in the Workspace)");
+  declareProperty(
+      new ArrayProperty<specid_t>("SpectrumList"),
+      "A comma-separated list of individual spectra to read.  Only used if\n"
+      "explicitly set.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Executes the algorithm
+ *  @throw std::out_of_range If a property is set to an invalid value for the
+ * input workspace
+ */
+void ExtractSpectra::exec() {
+  // Get the input workspace
+  m_inputWorkspace = getProperty("InputWorkspace");
+
+  eventW = boost::dynamic_pointer_cast<EventWorkspace>(m_inputWorkspace);
+  if (eventW != NULL) {
+    // Input workspace is an event workspace. Use the other exec method
+    this->execEvent();
+  } else {
+    // Otherwise it's a Workspace2D
+    this->execHistogram();
+  }
+}
+
+/// Execute the algorithm in case of a histogrammed data.
+void ExtractSpectra::execHistogram() {
+  m_histogram = m_inputWorkspace->isHistogramData();
+  // Check for common boundaries in input workspace
+  m_commonBoundaries = WorkspaceHelpers::commonBoundaries(m_inputWorkspace);
+
+  // Retrieve and validate the input properties
+  this->checkProperties();
+
+  // Create the output workspace
+  MatrixWorkspace_sptr outputWorkspace = WorkspaceFactory::Instance().create(
+    m_inputWorkspace, m_spectrumList.size(), m_maxX - m_minX,
+      m_maxX - m_minX - m_histogram);
+
+  // If this is a Workspace2D, get the spectra axes for copying in the spectraNo
+  // later
+  Axis *inAxis1(NULL), *outAxis1(NULL);
+  TextAxis *outTxtAxis(NULL);
+  if (m_inputWorkspace->axes() > 1) {
+    inAxis1 = m_inputWorkspace->getAxis(1);
+    outAxis1 = outputWorkspace->getAxis(1);
+    outTxtAxis = dynamic_cast<TextAxis *>(outAxis1);
```

Comment:
This was already there anyway, but I suspect that coverity will complain that outTxtAxis is used without checking the dynamic_cast outcome?